### PR TITLE
join lobby

### PR DESCRIPTION
**Pull request by LeviUniAcc**
Part of Gamify-IT/issues#375

pls review code and check on https://goulburn.informatik.uni-stuttgart.de:9443/minigames/towercrush/ if you can join a lobby

### CODE REVIEW

File: src/main/java/de/unistuttgart/towercrushbackend/data/websockets/Purpose.java [hunk 1]
```diff
@@ -1,5 +1,5 @@
 package de.unistuttgart.towercrushbackend.data.websockets;
 
 public enum Purpose {
-    CHAT_MESSAGE,JOIN_TEAM_MESSAGE
+    CHAT_MESSAGE, JOIN_TEAM_MESSAGE, JOIN_LOBBY_MESSAGE, DEVELOPER_MESSAGE
```
Comment (robojones) [inline src/main/java/de/unistuttgart/towercrushbackend/data/websockets/Purpose.java]:
```suggestion
    CHAT_MESSAGE,
    DEVELOPER_MESSAGE,
    JOIN_LOBBY_MESSAGE,
    JOIN_TEAM_MESSAGE
```
This list will become quite long, so I prefer having one value per line, in alphabetical order.

File: src/main/java/de/unistuttgart/towercrushbackend/controller/websockets/WebsocketController.java [hunk 2]
```diff
@@ -1,44 +1,54 @@
...
 
     @Autowired
     SimpMessagingTemplate simpMessagingTemplate;
-    final String lobbyDestination = "/topic/lobbies/";
+    @Autowired
+    LobbyManagerService lobbyManagerService;
+    @Autowired
+    WebsocketService websocketService;
+    final String newPlayerQueue = "/queue/private/messages";
+    final String lobbyTopic = "/topic/lobby/";
```
Comment (robojones) [inline src/main/java/de/unistuttgart/towercrushbackend/controller/websockets/WebsocketController.java]:
```suggestion
    static final String NEW_PLAYER_QUEUE = "/queue/private/messages";
    static final String LOBBY_TOPIC = "/topic/lobby/";
```
Naming convention for constants

File: src/main/java/de/unistuttgart/towercrushbackend/controller/websockets/WebsocketController.java [hunk 3]
```diff
@@ -1,44 +1,54 @@
...
+
+
+    @MessageMapping("/get/infos/on/join/{lobby}")
+    public void getInfosOnJoinLobby(@DestinationVariable final String lobby, @Header("simpSessionId") final String sessionId, final Principal user) throws JsonProcessingException {
+        log.info("Send lobby infos to newly joined player, sessionID: " + sessionId + ", User: " + user.getName());
+        final String sendTo = user.getName();
+        final Message joinLobbyMessage = new JoinLeaveLobbyMessage(lobbyManagerService.getLobby(lobby).getPlayerNames());
+        final MessageWrapper joinLobbyMessageWrapped = websocketService.wrapMessage(joinLobbyMessage, Purpose.JOIN_LOBBY_MESSAGE);
+        simpMessagingTemplate.convertAndSendToUser(sendTo, newPlayerQueue, joinLobbyMessageWrapped);
```
Comment (robojones) [inline src/main/java/de/unistuttgart/towercrushbackend/controller/websockets/WebsocketController.java]:
```suggestion
        simpMessagingTemplate.convertAndSendToUser(sendTo, WebsocketController.NEW_PLAYER_QUEUE, joinLobbyMessageWrapped);
```

File: src/main/java/de/unistuttgart/towercrushbackend/controller/components/WebsocketListener.java [hunk 4]
```diff
@@ -1,25 +1,179 @@
...
+    @Autowired
+    LobbyManagerService lobbyManagerService;
+    @Autowired
+    WebsocketService websocketService;
+    @Autowired
+    SimpMessagingTemplate simpMessagingTemplate;
+    ExecutorService executorService =
+        Executors.newFixedThreadPool(1);
+    final String lobbyTopic = "/topic/lobby/";
+    final String developerTopic = "/topic/developer/";
```
Comment (robojones) [inline src/main/java/de/unistuttgart/towercrushbackend/controller/components/WebsocketListener.java]:
```suggestion
    static final String LOBBY_TOPIC = "/topic/lobby/";
    static final String DEVELOPER_TOPIC = "/topic/developer/";
```
Naming convention for constants in Java

File: src/main/java/de/unistuttgart/towercrushbackend/controller/components/WebsocketListener.java [hunk 5]
```diff
@@ -1,25 +1,179 @@
...
+    /**
+     * This method sends all players in the lobby the new lobby list infos
+     *
+     * @param lobby lobby that gets informed
+     * @throws JsonProcessingException
+     */
+    private void sendLobbyBroadcastPlayerJoined(final String lobby) throws JsonProcessingException {
+        final Message joinLobbyMessage = new JoinLeaveLobbyMessage(lobbyManagerService.getLobby(lobby).getPlayerNames());
+        final MessageWrapper joinLobbyMessageWrapped = websocketService.wrapMessage(joinLobbyMessage, Purpose.JOIN_LOBBY_MESSAGE);
+        simpMessagingTemplate.convertAndSend(lobbyTopic + lobby, joinLobbyMessageWrapped);
```
Comment (robojones) [inline src/main/java/de/unistuttgart/towercrushbackend/controller/components/WebsocketListener.java]:
```suggestion
        simpMessagingTemplate.convertAndSend(WebsocketListener.LOBBY_TOPIC + lobby, joinLobbyMessageWrapped);
```

File: src/main/java/de/unistuttgart/towercrushbackend/controller/websockets/WebsocketController.java [hunk 6]
```diff
@@ -1,44 +1,54 @@
...
 
     @MessageMapping("/lobby/{lobby}/join/team/{team}/player/{player}")
     public void joinTeam(@DestinationVariable final String lobby, @DestinationVariable final String team, @DestinationVariable final String player) throws JsonProcessingException {
         log.info("player '{}' joined team '{}' in lobby '{}'", player, team, lobby);
         final JoinTeamMessage joinTeamMessage = new JoinTeamMessage(team, player);
-        final String jsonString = convertObjectToJson(joinTeamMessage);
-        final MessageWrapper joinTeamMessageWrapped = new MessageWrapper(jsonString, Purpose.JOIN_TEAM_MESSAGE);
-        simpMessagingTemplate.convertAndSend(lobbyDestination + lobby, joinTeamMessageWrapped);
+        final MessageWrapper joinTeamMessageWrapped = websocketService.wrapMessage(joinTeamMessage, Purpose.JOIN_TEAM_MESSAGE);
+        simpMessagingTemplate.convertAndSend(lobbyTopic + lobby, joinTeamMessageWrapped);
```
Comment (robojones) [inline src/main/java/de/unistuttgart/towercrushbackend/controller/websockets/WebsocketController.java]:
```suggestion
        simpMessagingTemplate.convertAndSend(WebsocketController.LOBBY_TOPIC + lobby, joinTeamMessageWrapped);
```

File: src/main/java/de/unistuttgart/towercrushbackend/controller/components/WebsocketListener.java [hunk 7]
```diff
@@ -1,25 +1,179 @@
...
+    }
+
+
+    /**
+     * As long as atleast one developer joined, a service sends ever x seconds a messsage to all developers to inform them of the current lobby status
+     */
+    private void handleDeveloperJoined() {
+        developerCount++;
+        if (developerTask != null) {
+            simpMessagingTemplate.convertAndSend(developerTopic,
```
Comment (robojones) [inline src/main/java/de/unistuttgart/towercrushbackend/controller/components/WebsocketListener.java]:
```suggestion
            simpMessagingTemplate.convertAndSend(WebsocketListener.DEVELOPER_TOPIC,
```

File: src/main/java/de/unistuttgart/towercrushbackend/controller/components/WebsocketListener.java [hunk 8]
```diff
@@ -1,25 +1,179 @@
...
+                "sending developer infos already started");
+            return;
+        }
+        simpMessagingTemplate.convertAndSend(developerTopic,
```
Comment (robojones) [inline src/main/java/de/unistuttgart/towercrushbackend/controller/components/WebsocketListener.java]:
```suggestion
        simpMessagingTemplate.convertAndSend(WebsocketListener.DEVELOPER_TOPIC,
```

File: src/main/java/de/unistuttgart/towercrushbackend/controller/components/WebsocketListener.java [hunk 9]
```diff
@@ -1,25 +1,179 @@
...
+            "Started developer infos");
+        developerTask = executorService.submit(() -> {
+            while (true) {
+                log.info("sending developer infos to single/multiple devs:" + developerCount);
+                final Message developerMessage = new DeveloperMessage(lobbyManagerService.getLobbies());
+                final MessageWrapper developerMessageWrapped = websocketService.wrapMessage(developerMessage, Purpose.DEVELOPER_MESSAGE);
+                simpMessagingTemplate.convertAndSend(developerTopic, developerMessageWrapped);
```
Comment (robojones) [inline src/main/java/de/unistuttgart/towercrushbackend/controller/components/WebsocketListener.java]:
```suggestion
                simpMessagingTemplate.convertAndSend(WebsocketListener.DEVELOPER_TOPIC, developerMessageWrapped);
```

File: src/main/java/de/unistuttgart/towercrushbackend/controller/components/WebsocketListener.java [hunk 10]
```diff
@@ -1,25 +1,179 @@
...
+    private void handleDeveloperDisconnected() {
+        developerCount--;
+        if (developerCount <= 0) {
+            log.info("no more devs online, stop infos");
+            developerCount = 0;
+            developerTask.cancel(true);
+            developerTask = null;
+        }
     }
 }
```
Comment (robojones) [inline src/main/java/de/unistuttgart/towercrushbackend/controller/components/WebsocketListener.java]:
```suggestion
}

```
add newline at end of file

**Review by robojones — APPROVED** [review]
Works

File: src/main/java/de/unistuttgart/towercrushbackend/data/websockets/Purpose.java (hunk 1 above)
Comment (LeviUniAcc) [reply src/main/java/de/unistuttgart/towercrushbackend/data/websockets/Purpose.java]: hm idk if the autoformatter will just change it again, but we can try

File: src/main/java/de/unistuttgart/towercrushbackend/controller/websockets/WebsocketController.java (hunk 2 above)
Comment (LeviUniAcc) [reply src/main/java/de/unistuttgart/towercrushbackend/controller/websockets/WebsocketController.java]: idk why but sure

File: src/main/java/de/unistuttgart/towercrushbackend/controller/websockets/WebsocketController.java (hunk 3 above)
Comment (LeviUniAcc) [reply src/main/java/de/unistuttgart/towercrushbackend/controller/websockets/WebsocketController.java]: sure, still dont know why

File: src/main/java/de/unistuttgart/towercrushbackend/controller/components/WebsocketListener.java (hunk 4 above)
Comment (LeviUniAcc) [reply src/main/java/de/unistuttgart/towercrushbackend/controller/components/WebsocketListener.java]: sure

File: src/main/java/de/unistuttgart/towercrushbackend/controller/components/WebsocketListener.java (hunk 5 above)
Comment (LeviUniAcc) [reply src/main/java/de/unistuttgart/towercrushbackend/controller/components/WebsocketListener.java]: sure

File: src/main/java/de/unistuttgart/towercrushbackend/controller/websockets/WebsocketController.java (hunk 6 above)
Comment (LeviUniAcc) [reply src/main/java/de/unistuttgart/towercrushbackend/controller/websockets/WebsocketController.java]: sure :)

File: src/main/java/de/unistuttgart/towercrushbackend/controller/components/WebsocketListener.java (hunk 8 above)
Comment (LeviUniAcc) [reply src/main/java/de/unistuttgart/towercrushbackend/controller/components/WebsocketListener.java]: sure

File: src/main/java/de/unistuttgart/towercrushbackend/controller/components/WebsocketListener.java (hunk 9 above)
Comment (LeviUniAcc) [reply src/main/java/de/unistuttgart/towercrushbackend/controller/components/WebsocketListener.java]: sure

File: src/main/java/de/unistuttgart/towercrushbackend/controller/components/WebsocketListener.java (hunk 10 above)
Comment (LeviUniAcc) [reply src/main/java/de/unistuttgart/towercrushbackend/controller/components/WebsocketListener.java]: yes

**Review by marlau97 — APPROVED** [review]

**Pull request merged**